provider/transip.go

```diff
@@ -0,0 +1,373 @@
+package provider
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/kubernetes-incubator/external-dns/endpoint"
+	"github.com/kubernetes-incubator/external-dns/plan"
+	log "github.com/sirupsen/logrus"
+	"github.com/transip/gotransip"
+	transip "github.com/transip/gotransip/domain"
+)
+
+const (
+	// 60 seconds is the current minimal TTL for TransIP and will replace unconfigured
+	// TTL's for Endpoints
+	transipMinimalValidTTL = 60
+)
+
+// TransIPProvider is an implementation of Provider for TransIP.
+type TransIPProvider struct {
+	client       gotransip.SOAPClient
+	domainFilter DomainFilter
+	dryRun       bool
+}
+
+// NewTransIPProvider initializes a new TransIP Provider.
+func NewTransIPProvider(accountName, privateKeyFile string, domainFilter DomainFilter, dryRun bool) (*TransIPProvider, error) {
+	// check given arguments
+	if accountName == "" {
+		return nil, errors.New("required --transip-account not set")
+	}
+
+	if privateKeyFile == "" {
+		return nil, errors.New("required --transip-keyfile not set")
+	}
+
+	var apiMode gotransip.APIMode
+	if dryRun {
+		apiMode = gotransip.APIModeReadOnly
+	} else {
+		apiMode = gotransip.APIModeReadWrite
+	}
+
+	// create new TransIP API client
+	c, err := gotransip.NewSOAPClient(gotransip.ClientConfig{
+		AccountName:    accountName,
+		PrivateKeyPath: privateKeyFile,
+		Mode:           apiMode,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("could not setup TransIP API client: %s", err.Error())
+	}
+
+	// return tipCloud struct
+	return &TransIPProvider{
+		client:       c,
+		domainFilter: domainFilter,
+		dryRun:       dryRun,
+	}, nil
+}
+
+// ApplyChanges applies a given set of changes in a given zone.
+func (p *TransIPProvider) ApplyChanges(changes *plan.Changes) error {
+	// build zonefinder with all our zones so we can use FindZone
+	// and a mapping of zones and their domain name
+	zones, err := p.fetchZones()
+	if err != nil {
+		return err
+	}
+
+	zoneNameMapper := zoneIDName{}
+	zonesByName := make(map[string]transip.Domain)
+	updatedZones := make(map[string]bool)
+	for _, zone := range zones {
+		// TransIP API doesn't expose a unique identifier for zones, other than than
+		// the domain name itself
+		zoneNameMapper.Add(zone.Name, zone.Name)
+		zonesByName[zone.Name] = zone
+	}
+
+	zoneForZoneName := func(name string, m zoneIDName, z map[string]transip.Domain) (zone transip.Domain, err error) {
```

Comment:
I wouldn't specify the var for the returning values, because it would be mixed compared to other declarations of your functions.